src/els_diagnostics_server.erl

```diff
@@ -0,0 +1,168 @@
+
+%%==============================================================================
+%% Exports
+%%==============================================================================
+
+-export([ start_link/2
+        ]).
+
+%% gen_server callbacks
+-export([ init/1
+        , handle_call/3
+        , handle_cast/2
+        , handle_info/2
+        ]).
+
+%% API
+-export([ on_save/1
+        , on_open/1
+        ]).
+
+%%==============================================================================
+%% Includes
+%%==============================================================================
+-include("erlang_ls.hrl").
+
+%%==============================================================================
+%% Macros
+%%==============================================================================
+
+%%==============================================================================
+%% Record Definitions
+%%==============================================================================
+-record(state, { server_name :: atom()
+               , uri :: uri()
+               , diagnostics :: term()
+               }).
+
+%%==============================================================================
+%% Type Definitions
+%%==============================================================================
+-type state() :: #state{}.
+
+%%==============================================================================
+%% API
+%%==============================================================================
+-spec start_link({atom(), uri()}, term()) -> {ok, pid()}.
+start_link({Id, Uri}, _Opts) ->
+  {ok, _} = gen_server:start_link({local, Id},
+                                  ?MODULE,
+                                  {Id, Uri},
+                                  []).
+
+%%==============================================================================
+%% gen_server callbacks
+%%==============================================================================
+-spec init({atom(), uri()}) -> {ok, state()}.
+init({Id, Uri}) ->
+  State = #state{ server_name = Id
+                , uri = Uri
+                },
+  {ok, State}.
+
+-spec handle_call(any(), any(), state()) -> {reply, any(), state()}.
+handle_call(get_diagnostics, _From, #state{diagnostics = Diagnostics} = S) ->
+  {reply, Diagnostics, S};
+handle_call(UnknownMessage, _From, S) ->
+  lager:info("Unknown handle_call! message: ~p", [UnknownMessage]),
+  {reply, ok, S}.
+
+-spec handle_cast(any(), state()) -> {noreply, state()}.
+handle_cast(run, #state{uri = Uri} = S) ->
+  Diagnostics =
+    lists:foldl(
+      fun(Mod, AccumulatedDiagnostics0) ->
+          Diagnostics = diagnostic(Mod, Uri),
+          AccumulatedDiagnostics = AccumulatedDiagnostics0 ++ Diagnostics,
+          send_notification(Uri, AccumulatedDiagnostics),
+          AccumulatedDiagnostics
+      end, [], [ els_compiler_diagnostics
+               , els_elvis_diagnostics
+               , els_dialyzer_diagnostics
+               ]),
```

Comment:
Eventually these should all run async, so we can report say compiler diagnostics as soon as they are ready, without having to wait for dialyzer to finish.

Reply:
All of them actually run async due to the gen_server:cast and each time a diagnostic becomes available, all diagnostics are sent off in the Publish Diagnostics message. If we don't send all, the other diagnostics are lost to the IDE.